src/Server/KeeperHTTPStorageHandler.cpp

```diff
@@ -0,0 +1,405 @@
+          * Poco::Timespan::MILLISECONDS)
+{
+}
+
+void KeeperHTTPStorageHandler::performZooKeeperRequest(
+    const Coordination::OpNum opnum, const std::string & storage_path, HTTPServerRequest & request, HTTPServerResponse & response)
+{
+    switch (opnum)
+    {
+        case Coordination::OpNum::Exists:
+            return performZooKeeperExistsRequest(storage_path, response);
+        case Coordination::OpNum::List:
+            return performZooKeeperListRequest(storage_path, response);
+        case Coordination::OpNum::Get:
+            return performZooKeeperGetRequest(storage_path, response);
+        case Coordination::OpNum::Set:
+            return performZooKeeperSetRequest(storage_path, request, response);
+        case Coordination::OpNum::Create:
+            return performZooKeeperCreateRequest(storage_path, request, response);
+        case Coordination::OpNum::Remove:
+            return performZooKeeperRemoveRequest(storage_path, request, response);
+        default:
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Trying to perform ZK request for unsupported OpNum. It's a bug.");
+    }
+}
+
+Coordination::ResponsePtr KeeperHTTPStorageHandler::awaitKeeperResponse(std::shared_ptr<Coordination::ZooKeeperRequest> request)
+{
+    auto response_promise = std::make_shared<std::promise<Coordination::ResponsePtr>>();
+    auto response_future = response_promise->get_future();
+
+    auto response_callback
+        = [response_promise](const Coordination::ResponsePtr & zk_response) mutable { response_promise->set_value(zk_response); };
+
+    const auto session_id = keeper_dispatcher->getSessionID(session_timeout.totalMilliseconds());
+    keeper_dispatcher->registerSession(session_id, response_callback);
+
+    try
+    {
+        if (!keeper_dispatcher->putRequest(std::move(request), session_id))
+            throw Exception(ErrorCodes::TIMEOUT_EXCEEDED, "Session {} already disconnected", session_id);
+
+        if (response_future.wait_for(std::chrono::milliseconds(operation_timeout.totalMilliseconds())) != std::future_status::ready)
+            throw Exception(ErrorCodes::TIMEOUT_EXCEEDED, "Operation timeout ({} ms) exceeded.", operation_timeout.totalMilliseconds());
+
+        auto result = response_future.get();
+
+        keeper_dispatcher->finishSession(session_id);
+        return result;
+    }
+    catch (...)
+    {
+        // it is obligatory to finish the session, as it affects num_alive_connections metric
+        keeper_dispatcher->finishSession(session_id);
+        throw;
+    }
+}
+
+void KeeperHTTPStorageHandler::performZooKeeperExistsRequest(const std::string & storage_path, HTTPServerResponse & response)
+{
+    Coordination::ZooKeeperExistsRequest zk_request;
+    zk_request.path = storage_path;
+
+    const auto result_ptr = awaitKeeperResponse(std::make_shared<Coordination::ZooKeeperExistsRequest>(std::move(zk_request)));
```

Comment:
I think you should be able to avoid getting sessions for read requests and expose `KeeperServer::putLocalReadRequest` from `KeeperDispatcher`